compiler/src/iree/compiler/API2/Embed.h

```diff
@@ -0,0 +1,257 @@
+
+// Sets a mnemonic phase name to run compilation to. Default is "end".
+// The meaning of this is pipeline specific. See IREEVMPipelinePhase
+// for the standard pipeline.
+IREE_EMBED_EXPORTED void ireeCompilerRunSetCompileToPhase(
+    struct iree_compiler_run_t *run, const char *phase);
+
+// Enables/disables verification of IR after each pass. Defaults to enabled.
+IREE_EMBED_EXPORTED void ireeCompilerRunSetVerifyIR(
+    struct iree_compiler_run_t *run, bool enable);
+
+// Runs a compilation pipeline.
+// Returns false and emits diagnostics on failure.
+enum iree_compiler_pipeline_t {
+  IREE_COMPILER_PIPELINE_STD = 0,
+  IREE_COMPILER_PIPELINE_HAL_EXECUTABLE = 1,
+};
+IREE_EMBED_EXPORTED bool ireeCompilerRunPipeline(
+    struct iree_compiler_run_t *run, enum iree_compiler_pipeline_t pipeline);
+
+// Outputs the current compiler state as textual IR to the output.
+IREE_EMBED_EXPORTED struct iree_compiler_error_t *ireeCompilerRunOutputIR(
+    struct iree_compiler_run_t *run, struct iree_compiler_output_t *output);
+
+// Assuming that the compiler has produced VM IR, converts it to bytecode
+// and outputs it. This is a valid next step after running the
+// IREE_COMPILER_PIPELINE_STD pipeline.
+IREE_EMBED_EXPORTED struct iree_compiler_error_t *
+ireeCompilerRunOutputVMBytecode(struct iree_compiler_run_t *run,
+                                struct iree_compiler_output_t *output);
+
+// Assuming that the compiler has produced VM IR, converts it to textual
+// C source and output it. This is a valid next step after running the
+// IREE_COMPILER_PIPELINE_STD pipeline.
+IREE_EMBED_EXPORTED struct iree_compiler_error_t *
+ireeCompilerRunOutputVMCSource(struct iree_compiler_run_t *run,
+                               struct iree_compiler_output_t *output);
+
+// Outputs the contents of a single HAL executable as binary data.
+// This is a valid next step after running the
+// IREE_COMPILER_PIPELINE_HAL_EXECUTABLE pipeline.
+IREE_EMBED_EXPORTED struct iree_compiler_error_t *
+ireeCompilerRunOutputHALExecutable(struct iree_compiler_run_t *run,
+                                   struct iree_compiler_output_t *output);
+
+//===----------------------------------------------------------------------===//
+// Sources.
+// Compilation sources are loaded into iree_compiler_source_t instances. These
+// instances reference an originating session and may contain a concrete
+// buffer of memory. Generally, when processing a source, its backing buffer
+// will be transferred out from under it (i.e. sources are single-use).
+// The actual source instance must be kept live until all processing is
+// completed as some methods of loading a source involve maintaining
+// references to backing resources.
+//===----------------------------------------------------------------------===//
+
+// Create and destroy source instances.
+IREE_EMBED_EXPORTED struct iree_compiler_source_t *ireeCompilerSourceCreate(
+    struct iree_compiler_session_t *session);
```

Comment:
is there any use case for empty sources or sources that have multiple calls on them? I think making ireeCompilerSourceOpenFile and ireeCompilerSourceWrapBuffer would be better as returning an iree_compiler_source_t (as an out param) instead - less likelihood of users misusing things (multiple OpenFile calls on the same source, no calls, etc). equivalent here is making static C++ functions for returning initialized classes instead of exposing constructors and mutator functions.

Reply:
I think I walked myself down a rabbit hole thinking about bytecode, temporary files, etc -- and did it this way with some of that in mind. But if I climb out and look around fresh, I think you may be right...

Reply:
I didn't know if you planned on allowing things like source file concatenation - like adding multiple sources to a single logical source or something. I could imagine just like we do source split -> sources[] also doing a sources[] concat -> source, but we still may want that as a static constructor-style call (`ireeCompilerSourceFromConcatenatedSources(sources[], &source)`).